java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java

```diff
@@ -817,11 +822,26 @@ private ProgramState stateWithException(ProgramState programState, BytecodeCFG.B
   }
 
   private void enqueueExceptionHandlers(Type exceptionType, ProgramState ps) {
-    programPosition.block.successors().stream()
-      .map(b -> (BytecodeCFG.Block) b)
-      .filter(BytecodeCFG.Block::isCatchBlock)
-      .filter(b -> isExceptionHandledByBlock(exceptionType, b))
-      .forEach(b -> enqueue(new ProgramPoint(b), ps));
+    List<BytecodeCFG.Block> blocksCatchingException = programPosition.block.successors().stream()
+        .map(b -> (BytecodeCFG.Block) b)
+        .filter(BytecodeCFG.Block::isCatchBlock)
+        .filter(b -> isExceptionHandledByBlock(exceptionType, b))
+        .collect(Collectors.toList());
+    if (!blocksCatchingException.isEmpty()) {
+      blocksCatchingException.forEach(b -> enqueue(new ProgramPoint(b), ps));
+      if (isCatchExhaustive(exceptionType, blocksCatchingException)) {
+        return;
+      }
+    }
+    // exception was not handled or was handled only partially, enqueue exit block with exceptional SV
+    ps.storeExitValue();
```

Comment:
Are we sure at this point that there is a Exceptional SV on top of the stack? Maybe we should enforce it with a precondition.